### PR TITLE
Unix.symlink: support forward slashes in symlink targets under Windows

### PR DESCRIPTION
**Pull request by nojb**
See [MPR#7564](https://caml.inria.fr/mantis/view.php?id=7564).

/cc @dra27

### CODE REVIEW

**Comment by fdopen** [comment]
`GetFullPathName` will transform a relative path `../foo` to an absolute path (like `C:\\bar\\foo`).
It's not longer the same link after this transformation. It will point to a different file, if your setup changes (bind mounts, re-mounts at a different location,etc.).

**Comment by nojb** [comment]
Thanks @fdopen, good point.  I guess I will just string replace every `/` by `\\` before calling the C stub then.

**Comment by dra27** [comment]
@nojb - yes, though be careful not to do that if the supplied path begins `\\?\` (i.e. should not be normalised in anyway).

Interestingly, Cmd's `mklink` command suffers in the same way. I'm half-wondering if this may be another of those little corner cases for why programs should always aim to use backslashes on Windows!

**Comment by nojb** [comment]
@dra27 Got it.  But that should be `\\?\`, right ?

**Comment by nojb** [comment]
I pushed a corrected fix.

**Comment by dra27** [comment]
Markdown got in the way - I think the edited version of my comment matches what you're saying!

**Comment by dra27** [comment]
This looks good to me, thank you - could you include a reference to that excellent series of blog posts (from the Mantis report) as a comment and also note that we can't use the normalisation functions in the Win32 API because of the need to keep relative paths?

For the testsuite, please could you test Unix.has_symlink and skip the test if it's not available, just so that we catch the case for users on older OSes which require elevation for symlinks. As an aside, I haven't (yet) tested where Unix.has_symlink works on newer Windows 10 builds in developer mode - are you testing on Windows 10 or Windows 7?

File: otherlibs/win32unix/unix.ml [hunk 1]
```diff
@@ -393,6 +393,12 @@ let mkfifo _name _perm = invalid_arg "Unix.mkfifo not implemented"
 external readlink : string -> string = "unix_readlink"
 external symlink_stub : bool -> string -> string -> unit = "unix_symlink"
 
+let normalize_path path =
```
Comment (dra27) [inline otherlibs/win32unix/unix.ml]: Given that it's not path normalisation, perhaps this could be `normalize_slashes` or `normalize_separator`?

**Comment by dra27** [comment]
A changes entry is needed too, please.

**Comment by dra27** [comment]
I'd welcome some other dev opinions on this change, given that `CreateSymbolicLink` does not appear to do normalisation (@damiendoligez, @xavierleroy, @alainfrisch?)

It seems a good idea to me, though as it's not strictly a bug, I'm guessing at this stage it should be 4.06.0, rather than 4.05.0?

**Comment by nojb** [comment]
Added reference, renamed to `normalize_slashes`, added a Changes entry (under "Bug fixes", not sure about that one), and added a test precheck to only run under Windows with symlink permissions.

(I tested on Win 10 in Developer Mode but I think that `Unix.symlink` still does not work without elevation because a special flag needs to be passed to `CreateSymbolicLink`, see https://blogs.windows.com/buildingapps/2016/12/02/symlinks-windows-10).

**Comment by damiendoligez** [comment]
@nojb @dra27 could you explain what problem is caused by forward slashes in symlinks?

**Comment by nojb** [comment]
@damiendoligez Currently, on Windows, if the file `foo\bar` exists and you call `Unix.symlink "foo/bar" "foobar"`, then a symbolic link named `foobar` is created pointing, not to the existing file `foo\bar`, but to a non-existing file called `foo/bar`.

**Comment by damiendoligez** [comment]
I'm lost. I thought windows accepted both `/` and `\` as separators, which would mean that opening `foo\bar` is the same as opening `foo/bar`, so if one exists the other must also exist. Or did Microsoft manage to mess this up for symlinks?

**Comment by dra27** [comment]
Alas, no - Windows internally does not accept forward slashes. However, because all the Win32 API functions perform normalisation before handing over to the kernel, for the most part you can pretend Windows does support them. It would appear that the contents of a symlink are not normalised on reading (I can't find that documented, though). I think this may be a consequence of where symlinks are resolved, which is after normalisation would have taken place.

**Comment by tomjridge** [comment]
I would argue against this sort of thing: the contents of a symlink is surely just a string, so fiddling about with the string the library user supplies in order to get some behaviour you think should apply in the future is not the right way to go and just makes things more confusing. The user should be responsible for this themselves (or add some function whose name makes things absolutely clear).

**Comment by tomjridge** [comment]
Just to follow up the last comment: people use symlinks for all sorts of things, not just pointing at files... at least, this is the case on Linux. Perhaps it is possible that your fix works correctly on Windows for all possible symlinks past present and future, but I wouldn't like to put any money on it.

**Comment by alainfrisch** [comment]
I think this is a good change, which will increase the likelihood of having programs designed primarily for Unix work under Windows.

AFAICT, the content of a symlink under Windows is really not just a "string" in the OCaml sense (i.e. a sequence of bytes).  It is rather a sequence of 16-bit integers.  The direction taken in #1200 is to use valid utf-8 encoded strings on the OCaml side.  This means that sequences that are invalid utf-16 cannot be created anyway.

I wouldn't be surprised that symlinks are also not just "blobs" on all Unices.  E.g. doesn't macOS perform some kind of Unicode normalization on symlinks as it does for filenames?

**Review by alainfrisch — APPROVED** [review]
LGTM

**Comment by alainfrisch** [comment]
I'll let @dra27 make the final call.

**Comment by tomjridge** [comment]
I don't use windows, and I think the views of windows users should take precedence in this case.

**Pull request merged**